packages/kbn-health-gateway-server/src/kibana/routes/status.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import https from 'https';
+import { URL } from 'url';
+import type { Request, ResponseToolkit } from '@hapi/hapi';
+import nodeFetch, { RequestInit, Response } from 'node-fetch';
+import type { IConfigService } from '@kbn/config';
+import type { Logger } from '@kbn/logging';
+import type { KibanaConfigType } from '../kibana_config';
+import { KibanaConfig } from '../kibana_config';
+
+const HTTPS = 'https:';
+
+const GATEWAY_STATUS_ROUTE = '/api/status';
+const KIBANA_STATUS_ROUTE = '/api/status';
+
+interface StatusRouteDependencies {
+  log: Logger;
+  config: IConfigService;
+}
+
+type Fetch = (path: string) => Promise<Response>;
+
+export function createStatusRoute({ config, log }: StatusRouteDependencies) {
+  const kibanaConfig = new KibanaConfig(config.atPathSync<KibanaConfigType>('kibana'));
+  const fetch = configureFetch(kibanaConfig);
+
+  return {
+    method: 'GET',
+    path: GATEWAY_STATUS_ROUTE,
+    handler: async (req: Request, h: ResponseToolkit) => {
+      const responses = await fetchKibanaStatuses({ fetch, kibanaConfig, log });
+      const { body, statusCode } = mergeStatusResponses(responses);
+      return h.response(body).type('application/json').code(statusCode);
+    },
+  };
+}
+
+async function fetchKibanaStatuses({
+  fetch,
+  kibanaConfig,
+  log,
+}: {
+  fetch: Fetch;
+  kibanaConfig: KibanaConfig;
+  log: Logger;
+}) {
+  const requests = await Promise.allSettled(
+    kibanaConfig.hosts.map(async (host) => {
+      log.debug(`Fetching response from ${host}${KIBANA_STATUS_ROUTE}`);
+      const response = fetch(`${host}${KIBANA_STATUS_ROUTE}`).then((res) => res.json());
+      return response;
+    })
+  );
+
+  return requests.map((r, i) => {
+    if (r.status === 'rejected') {
+      log.error(`Unable to retrieve status from ${kibanaConfig.hosts[i]}${KIBANA_STATUS_ROUTE}`);
+    } else {
+      log.info(
+        `Got response from ${kibanaConfig.hosts[i]}${KIBANA_STATUS_ROUTE}: ${JSON.stringify(
+          r.value.status?.overall ? r.value.status.overall : r.value
+        )}`
+      );
+    }
+    return r;
+  });
+}
+
+function mergeStatusResponses(
+  responses: Array<PromiseFulfilledResult<Response> | PromiseRejectedResult>
+) {
+  let statusCode = 200;
+  for (const response of responses) {
+    if (response.status === 'rejected') {
```

Comment:
I think we should change this. My understanding is that status will be `rejected` when there is an issue in the SSL verification.
I think we should check explicitly the status code is one that we consider healthy (so either `status.ok` or `=== 401`) and return 503 in all other cases

Reply:
You're right, we need to be checking `response.value.ok` instead, because the `fetch` API will only reject on a network error and not on a 4xx or 5xx response.

This was an oversight on my part -- @dokmic something to look at when you are revisiting the logic here.